### ISSUE TITLE
Here is what I would propose for our startup story in v6 

### ISSUE THREAD

**Issue by johnsimons** [Particular/NServiceBus#3129]
I may have missed some scenarios

```c#
// Using Disposable pattern
using (var endpoint = await Endpoint.Start(config))
{
    using (var context = endpoint.CreateBusContext())
    {
        context.Send(new MyMessage());
    }
}

// Using async
var endpoint = await Endpoint.Start(config);

using (var context = endpoint.CreateBusContext())
{
    context.Send(new MyMessage());
}

await endpoint.Stop();

// From a console
using (var endpoint = await Endpoint.Start(config))
{
    using (var context = endpoint.CreateBusContext())
    {
        // Do initial sends
        context.Send(new MyMessage());
    }

    Console.WriteLine("Press any key to exit");
    Console.ReadKey();
}

// From a HttpApplication
public class MvcApplication : System.Web.HttpApplication
{
    private IEndpointInstance endpoint;

    protected void Application_Start()
    {
        endpoint = Endpoint.Start(config).Wait();

        using (var context = endpoint.CreateBusContext())
        {
            context.Send(new MyMessage());
        }
    }

    protected void Application_Stop()
    {
        endpoint.Stop().Wait();
    }
}

// From OWIN startup class
public partial class Startup
{
    private IEndpointInstance endpoint;

    public void Configuration(IAppBuilder app)
    {
        var context = new OwinContext(app.Properties);
        var token = context.Get<CancellationToken>("host.OnAppDisposing");

        endpoint = Endpoint.Start(config).Wait();

        using (var context = endpoint.CreateBusContext())
        {
            context.Send(new MyMessage());
        }

        if (token != CancellationToken.None)
        {
            token.Register(() =>
            {
                endpoint.Stop().Wait();
            });
        }
    }
}
```

So in summary I made `IEndpointInstance` disposable and the Dispose method just executes `Stop().Wait()`
I also made `IBusContext` disposable, this one is just for convenience if we need to dispose of anything later on, so currently is a noop

**Comment by johnsimons** [comment]
@Particular/nservicebus thougths ?

**Comment by danielmarbach** [comment]
@johnsimons the problem is: it is an antipattern :(

Synchronous wrapper which calls async method in a blocking nature
Masks the original exception
Potential source of deadlocks

More about IDisposable and async

http://blog.stephencleary.com/2013/03/async-oop-6-disposal.html

I'll do so more investigation.

**Comment by SimonCropp** [comment]
considering the desire here is to enable the using pattern i actually see no justification for IDisposable.

The reason is any real world use does not allow a using statement. eg windows service, webapi, mvc etc. All of these thing the context that you start the bus and stop the bus is different. so u cant use a using. I know all the samples use a using but they are not "real world" scenarios.

so I am -1 doing any kind of blocking hacks to enable syntactic sugar that the majority of people have no use for.

**Comment by johnsimons** [comment]
I am super confused but is a user wants to start and stop a bus in HttpApplication how do they go about that ? Don't they need to call Wait on the Start/Stop tasks ?
```c#
public class MvcApplication : System.Web.HttpApplication
{
    private IEndpointInstance endpoint;

    protected void Application_Start()
    {
        endpoint = Endpoint.Start(config).Wait();

        using (var context = endpoint.CreateBusContext())
        {
            context.Send(new MyMessage());
        }
    }

    protected void Application_Stop()
    {
        endpoint.Stop().Wait();
    }
}
```

Or am I looking at this wrong ?

**Comment by johnsimons** [comment]
Or in a OWIN Startup ?
```c#
public partial class Startup
{
    private IEndpointInstance endpoint;

    public void Configuration(IAppBuilder app)
    {
        var context = new OwinContext(app.Properties);
        var token = context.Get<CancellationToken>("host.OnAppDisposing");

        endpoint = Endpoint.Start(config).Wait();

        using (var context = endpoint.CreateBusContext())
        {
            context.Send(new MyMessage());
        }

        if (token != CancellationToken.None)
        {
            token.Register(() =>
            {
                endpoint.Stop().Wait();
            });
        }
    }
}
```

**Comment by johnsimons** [comment]
From my limited understanding of using the `await`, it can only be called on an async method, or am I missing something?

**Comment by johnsimons** [comment]
Same for when running under a windows service:
```c#
partial class Service1 : ServiceBase
{
    public Service1()
    {
        InitializeComponent();
    }

    protected override void OnStart(string[] args)
    {
        // TODO: Add code here to start your service.
    }

    protected override void OnStop()
    {
        // TODO: Add code here to perform any tear-down necessary to stop your service.
    }
}
```

**Comment by johnsimons** [comment]
@SimonCropp how about the `CreateBusContext` being `IDisposable`, is that a YAGNI too ? Or do we see benefits there ?

**Comment by danielmarbach** [comment]
@johnsimons there is nothing to dispose in the `BusContext`.

**Comment by danielmarbach** [comment]
@johnsimons The difference is

If we implement the blocking wait inside our library code there is no way the user can control this.

If we return a Task like we do know the user can decide at which level of the call stack they interrupt the asynchronous flow and wether they use `AsyncPump`, `GetAwaiter().GetResult()` or `Wait()` whic of course wraps the original exception in an `AggregateException`

**Comment by johnsimons** [comment]
@danielmarbach the only place we would be blocking inside our lib in my proposal in the `Dispose()` methods, does that make sense ?

**Comment by timbussmann** [comment]
```
// From a console
using (var endpoint = await Endpoint.Start(config))
{
```
you can't use await here

What about a fire & forget stop when calling dispose?

**Comment by danielmarbach** [comment]
@johnsimons we all agreed that we won't do this. Since we have no proper way to implement it without potentially running into dead locks. I'm closing it.
